Add support for Lidar sensor

bmalhigh opened the pull request:
Add Lidar sensor in AirLib object model as well as Unreal
implementation using ray-casting. Add python client example.

Future work:
    - Enable configuration of lidar parameters
    - Enable visualization of point-cloud on client side

Resolves: #621

bmalhigh committed:
Add support for Lidar sensor

Add Lidar sensor in AirLib object model as well as Unreal
implementation using ray-casting. Add python client example.

Future work:
    - Enable configuration of lidar parameters
    - Enable visualization of point-cloud on client side

Resolves: #621

bmalhigh committed:
Add support for Lidar sensor

Add Lidar sensor in AirLib object model as well as Unreal
implementation using ray-casting. Add python client examples.

Future work:
    - Enable configuration of lidar parameters
    - Enable visualization of point-cloud on client side

Resolves: #621

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/AirSim?pullRequest=1424) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/AirSim?pullRequest=1424) before we can accept your contribution.<br/><br/>:x: bmalhigh [sign now](https://cla.opensource.microsoft.com/Microsoft/AirSim?pullRequest=1424)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/AirSim?pullRequest=1424) it.</sub>

bmalhigh committed:
Fix lidar support for cars

Reference: #621

bmalhigh committed:
Merge branch 'feature/lidar' of https://github.com/bmalhigh/AirSim into feature/lidar

sytelus commented on Unreal/Plugins/AirSim/Source/UnrealSensors/UnrealLidarSensor.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    msr::airlib::LidarSimpleParams params = getParams();
+
+    const auto numberOfLasers = params.number_of_channels;
```
would be good to use _ for local vars in this method for consistency.

bmalhigh replied:
Yep. Hopefully fixed everywhere now.

sytelus commented on Unreal/Plugins/AirSim/Source/UnrealSensors/UnrealLidarSensor.cpp:
```diff
@@ -0,0 +1,144 @@
+        if (false)
+        {
+            // Debug code.
+            // don't enable this for multirotors since drawing a debug point
```
You can use `UAirBlueprintLib::IsInGameThread()` to check if you are allowed to do draw calls

bmalhigh replied:
Fixed

sytelus commented on Unreal/Plugins/AirSim/Source/UnrealSensors/UnrealLidarSensor.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    // calculate end position (params.range is expected to be in meters)
+    constexpr float TO_CENTIMETERS = 1e+2f;
+    FVector end = (params.range * TO_CENTIMETERS) * UKismetMathLibrary::GetForwardVector(resultRot) + start;
```
I'm wondering if this computation can be bit cleaned up. We have 3 coordinate frames: world frame, vehicle frame, lidar frame. The usual way to do computation is to start with inner most frame and then ultimately get the vector in world frame. Also, it's often useful to do all computation in just one system (for example, either NED or UE). Usually we just keep everything in NED system and convert to UE system only when UE API call is needed. This keeps code portable to other engines (we have Unity port in works).

Below is just pseudo code. The suffix _w, _b and _l are used for world, body and lidar frames.

We have:
```
ray_h_l, ray_v_l = horizontal and vertical angles for ray in lidar frame
lidar_pose_b = lidar pose relative to vehicle body
vehicle_pose_w = vehicle pose relative to world frame

//get ray quaternion in lidar frame (angles must be in radians)
ray_q_l = VecorMath::toQuaternion(ray_v_l, 0, ray_h_l)
//get ray quaternion in body frame
ray_q_b = VectorMath::rotateQuaternion(ray_q_l, lidar_pose_b.orientation)
//get ray quaternion in world frame
ray_q_w = VectorMath::rotateQuaternion(ray_q_b, vehicle_pose_w.orientation)
//get ray vector
ray_w = VectorMath::rotateVector(VectorMath::front() * max_distance, ray_q_w) 
   + lidar_pose_b.position + vehicle_pose_w.position
```

sytelus commented on Unreal/Plugins/AirSim/Source/Vehicles/Multirotor/MultirotorPawnSimApi.cpp:
```diff
@@ -49,6 +49,8 @@ void MultirotorPawnSimApi::pawnTick(float dt)
     unused(dt);
     //calls to update* are handled by physics engine and in SimModeWorldBase
+
+    drawLidarDebugPoints();
```
If we move this out of PawnSimAPi to SimModeBase, this won't needed because visualization code will be agnostic to vehicle type.

bmalhigh replied:
Fixed.

sytelus commented on AirLib/include/vehicles/car/api/CarApiBase.hpp:
```diff
@@ -63,21 +66,105 @@ class CarApiBase : public VehicleApiBase  {
+        return sensors_;
+    }
+
+    void initialize(std::shared_ptr<SensorFactory> sensor_factory, const Kinematics::State& state, const Environment& environment)
```
Just wondering if we should just add this to constructor? That way someone using this class won't forget to call initialize.

bmalhigh replied:
Done. It required to pass the params through the intermediate base class; so not very clean looking... but I made the change for now.

sytelus reviewed (commented):
Looks really good! Most of the comments are just minor nitpicks. I think we are very close to merging these changes.

KerryMoffittRtn commented:
Oh, this is great; thanks for doing it!  One thought: consider using the existing sim depth sensor, and resampling from the fixed-width / fixed-height sample deltas that one gets from a Kinect / flash lidar / existing sim-depth-sensor, instead of simulating each individual laser fiting.  That would presumably provide a giant speedup, since you can use the existing (GPU-based) render pipeline, vs. performing all the collision calculations on the CPU.  Cheers!

bmalhigh commented on AirLib/include/api/RpcLibAdapatorsBase.hpp:
```diff
@@ -461,6 +461,40 @@ class RpcLibAdapatorsBase {
+
+    struct LidarData {
+
+        // header
```
My intent here was that "header" and "data" were more like section headers... with the assumption that there will be multiple fields in header section along the way... but fixed for now.

bmalhigh commented on AirLib/include/api/VehicleApiBase.hpp:
```diff
@@ -94,6 +96,30 @@ class VehicleApiBase : public UpdatableObject {
+    virtual LidarData getLidarData(const std::string& lidar_name) const
+    {
+        // TODO: find the appropriate Lidar based on the name
+        const LidarBase* lidar = static_cast<const LidarBase*>(getSensors().getByType(SensorBase::SensorType::Lidar));
```
AFAIK, no. SensorCollection::getByType(...) returns only one SensorBase back. For now fixing it by using the combination of size() and using index in getByType().

bmalhigh commented on AirLib/include/api/VehicleApiBase.hpp:
```diff
@@ -94,6 +96,30 @@ class VehicleApiBase : public UpdatableObject {
+            throw VehicleControllerException("No matching lidar found on the vehicle");
+        }
+
+        LidarData lidarData;
```
Fixed

bmalhigh commented on AirLib/include/common/CommonStructs.hpp:
```diff
@@ -295,5 +295,14 @@ struct RCData {
+struct LidarData {
+
+    TTimePoint time_stamp = 0;
+    vector<float> point_cloud;
```
I guess we are not being consistent with real_T vs. float... I see other places in AirLib where we are using vector<float>... for example -- JoyStick::axes, ImageResponse::image_data_float... but I changed the LidarData::point_cloud to use real_T

bmalhigh commented on AirLib/include/sensors/lidar/LidarBase.hpp:
```diff
@@ -0,0 +1,56 @@
+
+#include "sensors/SensorBase.hpp"
+
+namespace msr {
```
Fixed

bmalhigh commented on AirLib/include/vehicles/multirotor/MultiRotorParams.hpp:
```diff
@@ -32,6 +32,7 @@ class MultiRotorParams {
         bool gps = true;
         bool barometer = true;
         bool distance = false; //this causes ray casts so disabled by default
+        bool lidar = true;     //this causes ray casts so disabled by default: TODO
```
Yes, that was the intent. Fixed.

bmalhigh commented on Unreal/Plugins/AirSim/Source/UnrealSensors/UnrealLidarSensor.h:
```diff
@@ -0,0 +1,37 @@
+    AActor* actor_;
+    const NedTransform* ned_transform_;
+
+    TArray<float> laser_angles_;
```
Fixed.

bmalhigh commented on Unreal/Plugins/AirSim/Source/PawnSimApi.cpp:
```diff
@@ -10,6 +10,8 @@
 #include "NedTransform.h"
 #include "common/EarthUtils.hpp"
 
+#include "DrawDebugHelpers.h"
```
Moved to SimModeBase.

bmalhigh commented on AirLib/include/sensors/lidar/LidarSimple.hpp:
```diff
@@ -0,0 +1,99 @@
+            virtual ~LidarSimple() = default;
+
+        protected:
+            virtual vector<real_T> getPointCloud(const Pose& pose, float deltaTime) = 0;
```
Done.

bmalhigh commented on AirLib/include/api/VehicleApiBase.hpp:
```diff
@@ -94,6 +96,30 @@ class VehicleApiBase : public UpdatableObject {
+        LidarBase::Output output = lidar->getOutput();
+        
+        lidarData.time_stamp = output.time_stamp;
+        lidarData.point_cloud = std::vector<float>(output.point_cloud);
```
Done

bmalhigh committed:
Incorporate PR review feedback on Lidar pull-request.

1 - Change coordinate math in UnrealLidarSensor to use AirLib vector math.
2 - Move draw support to SimMode/SimModeBase.
3 - Fix coding style/convention discrepancies.

bmalhigh commented:
> Oh, this is great; thanks for doing it! One thought: instead of simulating each individual laser firing, consider using the existing sim depth sensor, and resampling from the fixed-width / fixed-height sample deltas that one gets from a Kinect / flash lidar / existing sim-depth-sensor to the fixed-angle deltas one gets with a spinning lidar like a traditional Velodyne. That would presumably provide a giant speedup, since you could use the existing (GPU-based) render pipeline, vs. performing all the collision calculations on the CPU. Cheers!

@KerryMoffittRtn -- I agree with the goal to reduce the number of lasers fired or the number of calcs done. However, I am unable to understand the solution that you are proposing above. Apologies for being dense. Can you clarify the above resampling approach in bit more detail?

Cheers.

KerryMoffittRtn commented:
Sure!  As I understand your approach, you cast a ray per laser firing, and test for collision in the world geometry.  This seems sound, and will presumably provide the functionality you desire, though it means asking Unreal to navigate the geometry of the scene to perform collision detection (presumably on the CPU) for every ray cast.  To understand the alternative approach I am describing, first consider the existing depth capture mechanism (see ImageType::DepthVis, USceneCaptureComponent2D, and the picture-in-picture that the '1' key toggles).  It works a lot like the standard scene renderer (heavily optimized, and taking full advantage of the parallelism offered by the GPU) except that it yields depth information instead of color information.  Now this is super fast, and outputs a 2D array of depth samples (lots of samples, i.e.) at once, _but_ they're not geometrically structured the way you want: the rays all pass through a focal plane, on which we have a fixed width and height between adjacent samples.  What you want for a spinning lidar is fixed _angle_ between adjacent samples.  So you would need to reorganize the data by resampling, and the cost of this resampling (though it may not be terrible) would need to be weighed against the fast performance of the initial depth sampling.  (You'd also need to collect samples on multiple planes, of course - at least three - to cover all 360 degrees of a circle.)  This is a little tricky, I realize.  Happy to chat more if I'm still not making sense.  :)

TritonSailor commented:
This is a very useful feature. If performance needs to be optimized, it would probably be helpful to ask in the unreal forums how to most efficiently implement this. Also, you might find some useful code in the Nvidia Waveworks project. It uses ray casting (on the gpu) to determine bouyancy of actors at various locations on the map.

anphunl commented on AirLib/include/api/VehicleApiBase.hpp:
```diff
@@ -11,6 +11,8 @@
 #include "common/CommonStructs.hpp"
 #include "common/ImageCaptureBase.hpp"
+#include "sensors/SensorCollection.hpp"
+#include "sensors/Lidar/LidarBase.hpp"
```
Should be #include "sensors/lidar/LidarBase.hpp", shouldn't it?

SamHS-OCI replied:
I think you're right, I was having build errors, replaced the suggested line and now it is working. Thanks.

bmalhigh replied:
Ah, Linux issue... apologies... fixing now.

andre-nguyen commented:
@KerryMoffittRtn Is that how Gazebo does it? I think I saw something like that for the velodyne simulator in Gazebo and I was a bit confused as to why we would do it using cameras. 

Is there no way of manually raytracing on the gpu without going through the rendering pipeline?
